planner/optimize.go

```diff
@@ -264,6 +293,22 @@ func optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		return nil, nil, 0, err
 	}
 
+	if !sctx.GetSessionVars().InRestrictedSQL {
+		readOnly, err := sctx.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.TiDBRestrictedReadOnly)
+		if err != nil {
+			return nil, nil, 0, err
+		}
+		if readOnly == variable.On {
```

Comment:
This value is technically read from a table which could include user-supplied data, so it could be `ON`/`on`/`On` (although if users are doing things with `SET x=y` it never will because of validation). In other cases in the server we use `variable.TiDBOptOn(readOnly)` to be sure.

(Assuming you change this to an atomic, this comment becomes moot.)